### PR TITLE
Initial README

### PR DESCRIPTION
**Pull request by JensRantil**
Information about:
- What this project is about.
- Its goals.
- Its values.
- A basic vision of where this project should be going.

### CODE REVIEW

**Comment by JensRantil** [comment]
@eripe970 Thoughts?

**Comment by JensRantil** [comment]
Additional thoughts (that might need to be put in writing in README, too):
- Two types of plugins:
  - Those that stores the migration states for each migration. Can be stored in plain file, in database, zookeeper whatnot. I've called these _repositories_ for lack of a better word.
  - Those that execute migrations, be it running SQL executions, a script etc. For lack of a better word I've called these _migrators_.
- I'm thinking of a plugin system like https://github.com/hashicorp/go-plugin.
  - Pros:
    - Will allow _anyone_ to write their own plugins.
    - Simply drop in a plugin binary and it will automatically discovered by `any-migrate`.
    - Doesn't require recompiling `any-migrate` to get plugins in.
    - Doesn't require `cgo` or any dynamically linked library and everything that comes with that.
  - Cons:
    - Handling backward compatability etc.

Discussion points:
- Are there too many states? Too few? I think testing migrations (if possible) and supporting reverts is important.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+---------------
+
+First, [install Go](https://golang.org/dl/) and then set the `GOPATH`
+environment variable to a directory where you'd like to install the binaries.
+You can then install the binaries needed for this example by issuing
+
+```bash
+$ go install github.com/any-migrate/any-migrate
+$ go install github.com/any-migrate/file-repository
+$ go install github.com/any-migrate/mysql-migrator
```
Comment (JensRantil) [inline README.md]: I think splitting this into separate repositories probably isn't a good idea in the beginning. Better to stabilize integration points (that is, interfaces) before splitting out into separate repositories.

**Comment by eripe970** [comment]
Like the idea with different plugins! I also think there could be a need for a reporting plugin (to graphite for example) if you would like to store your migration state in mysql or text file but still would like to push information to graphite when you did your migrations. 

Some other things that crossed my mind. 
- Do we want to support migrations of migration states to different repositories? I think so. I'm thinking that users will start playing around with a textfile and then would like to move up to a database in an easy way.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+complex than that and you want to execute a script. The script could clean up
+your database, execute `pt-schema-migrate` (for MySQL), do complex migrations
+of users, populate complex data, or other one-off tasks you'd like to execute.
+`any-migrate` is a tool aiming to support custom data migrations.
+
+`any-migrate` will have the following features:
+
+ * It keeps track of which migrations has been executed on a system.
+ * Fails as early as possible.
+ * Ask before assuming something. Explicitness is important. No magic.
```
Comment (eripe970) [inline README.md:32]: Agree on this. Think it is really important.  Think there should be an option to say "Y/N" before each migration.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+```yaml
+# my.config
+repository:
+  type: file
+  path: ./my.repository
+migrations:
+  path: ./my.changes
+sources:
+  mysql:
+    migrator: mysql-migrator
```
Comment (eripe970) [inline README.md:136]: One thing that I really hate is when something works on development/staging but doesn't work on production. I think that it is important that this software tries to minimze this as much as possible.  

Two things that I think about that often differs between environments.
- Versions of software (for example mysql versions)
- Access (for example granted rights in mysql)

Access can be a bit tricky but I could see a point in having a 1A and 1B version of a migration where A version is applied to versions running version `A` and B is applied to versions running `B`. 

Specifying a migration on the form will make it easier to check access compared to a hard coded sql script. 

```
action:
  type: rename table
    from: ussers
    to: users
```

File: README.md (hunk 2 above)
Comment (JensRantil) [inline README.md:32]: True. At the same time it must be possible to fully automate migrations.

File: README.md (hunk 3 above)
Comment (JensRantil) [inline README.md:136]: > One thing that I really hate is when something works on development/staging but doesn't work on production. I think that it is important that this software tries to minimze this as much as possible.

True. That said, it _is_ bad practise to have different versions and nobody should blaim this tool for their error. But, in general, I can agree that we should try to minimize it.

Regarding your `action` example above, I think the "high level" actions (such as "rename table") both will simplify checking access as well as as handling different versions. That said, I think it's up to a migrator plugin to handle check version/access, not the core "any-migrate".

Thoughts?

**Comment by JensRantil** [comment]
> Like the idea with different plugins! I also think there could be a need for a reporting plugin (to graphite for example) if you would like to store your migration state in mysql or text file but still would like to push information to graphite when you did your migrations.

I think it's a good idea, but that's maybe something that we can leave out until version 0.2? I created 0.1 and 0.2 milestones in the issues tab and assigned #6 to milestone 0.2.

This also made me think of logging, see #7.

> Do we want to support migrations of migration states to different repositories? I think so. I'm thinking that users will start playing around with a textfile and then would like to move up to a database in an easy way.

I think that would be great! However, however I propose we leave that for milestone 0.2. I've created an issue for it [here](#9).

File: README.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+```bash
+$ go install github.com/any-migrate/any-migrate
+$ go install github.com/any-migrate/file-repository
+$ go install github.com/any-migrate/mysql-migrator
+```
+
+`any-migrate` is the command line utility that you'll use to interact with
+any-migrate. A "repository" is a plugin which is used for keeping track of
+which migrations have been applied. The "file" stores this informtion in a
```
Comment (JensRantil) [inline README.md]: Typo.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+plain file. A migrator is a plugin which applies changes. The "mysql" plugin
```
Comment (JensRantil) [inline README.md]: "migrator" in italics.

File: README.md (hunk 4 above)
Comment (JensRantil) [inline README.md]: "file" in italics.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+You can then install the binaries needed for this example by issuing
+
+```bash
+$ go install github.com/any-migrate/any-migrate
+$ go install github.com/any-migrate/file-repository
+$ go install github.com/any-migrate/mysql-migrator
+```
+
+`any-migrate` is the command line utility that you'll use to interact with
+any-migrate. A "repository" is a plugin which is used for keeping track of
```
Comment (JensRantil) [inline README.md]: "repository" in italics.

File: README.md (hunk 2 above)
Comment (eripe970) [inline README.md:32]: > True. At the same time it must be possible to fully automate migrations.

Agree!

File: README.md (hunk 1 above)
Comment (eripe970) [inline README.md]: > I think splitting this into separate repositories probably isn't a good idea in the beginning. Better to stabilize integration points (that is, interfaces) before splitting out into separate repositories.

Agree on that as well!

File: README.md (hunk 3 above)
Comment (eripe970) [inline README.md:136]: Yeah, agree on that it is nice to use as much "high level" actions as possible but still support "low level". You don't want to configure a whole table creation in a new tool, just copy and paste the sql-create statement.

**Pull request merged**